nighthawk/source/client/benchmark_client_impl.cc

```diff
@@ -0,0 +1,230 @@
+    // integration test because of a runtime/tls conflict with the integration test server, when
+    // the message gets validated. Ideally we'd just re-use Tls::Upstream::createTransportFactory()
+
+    auto client_config =
+        std::make_unique<Envoy::Extensions::TransportSockets::Tls::ClientContextConfigImpl>(
+            dynamic_cast<const envoy::api::v2::auth::UpstreamTlsContext&>(*message),
+            *transport_socket_factory_context_);
+    socket_factory =
+        std::make_unique<Envoy::Extensions::TransportSockets::Tls::ClientSslSocketFactory>(
+            std::move(client_config), transport_socket_factory_context_->sslContextManager(),
+            transport_socket_factory_context_->statsScope());
+  } else {
+    socket_factory = std::make_unique<Envoy::Network::RawBufferSocketFactory>();
+  };
+
+  cluster_ = std::make_unique<Envoy::Upstream::ClusterInfoImpl>(
+      cluster_config, bind_config, runtime, std::move(socket_factory),
+      store_->createScope("client."), false /*added_via_api*/);
+
+  auto host = std::shared_ptr<Envoy::Upstream::Host>{new Envoy::Upstream::HostImpl(
+      cluster_, uri_->host_and_port(), target_address_,
+      envoy::api::v2::core::Metadata::default_instance(), 1 /* weight */,
+      envoy::api::v2::core::Locality(),
+      envoy::api::v2::endpoint::Endpoint::HealthCheckConfig::default_instance(), 0,
+      envoy::api::v2::core::HealthStatus::HEALTHY)};
+
+  Envoy::Network::ConnectionSocket::OptionsSharedPtr options =
+      std::make_shared<Envoy::Network::ConnectionSocket::Options>();
+  if (use_h2_) {
+    pool_ = std::make_unique<Envoy::Http::Http2::ProdConnPoolImpl>(
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  } else {
+    pool_ = std::make_unique<Envoy::Http::Http1::ProdConnPoolImpl>(
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  }
+  return true;
+}
+
+void BenchmarkClientHttpImpl::terminate() { pool_.reset(); }
+
+StatisticPtrMap BenchmarkClientHttpImpl::statistics() const {
+  StatisticPtrMap statistics;
+  statistics[connect_statistic_->id()] = connect_statistic_.get();
+  statistics[response_statistic_->id()] = response_statistic_.get();
+  return statistics;
+};
+
+bool BenchmarkClientHttpImpl::tryStartOne(std::function<void()> caller_completion_callback) {
+  ASSERT(target_address_.get());
+  if (!cluster_->resourceManager(Envoy::Upstream::ResourcePriority::Default)
+           .pendingRequests()
+           .canCreate() ||
+      // In closed loop mode we want to be able to control the pacing as exactly as possible.
+      // In open-loop mode we probably want to skip this.
+      // NOTE(oschaaf): We can't consistently rely on resourceManager()::requests() because that
+      // isn't used for h/1 (it is used in tcp and h2 though).
+      ((requests_initiated_ - requests_completed_) >= connection_limit_)) {
+    return false;
+  }
+
+  auto stream_decoder = new StreamDecoder(
```

Comment:
@htuch re: "is it possible that the allocating entity can also do cleanup via a smartptr, or do we need to self-destruct"

I'm not sure, as I understand it `StreamDecoder` relies on Envoy to perform the right callbacks, which will at some point drive it into its terminal state (and `StreamDecoder` is responsible for correctly interpreting these callbacks and determine when it has served its purpose). 

I wonder, wouldn't it be nice if somehow we could optionally transfer or share ownership of the decoder to/with Envoy in the call to `newStream` below? Then we could choose here: we transfer ownership (no pooling), or we share ownership and re-use it  (e.g. recycle to a decoder pool) once Envoy is done with it.

Reply:
Resolving this as per out-of-band discussion